### PR TITLE
Portuguese translation for alternatives.md

### PR DESCRIPTION
**Pull request by Serrones**

### CODE REVIEW

File: docs/pt/docs/alternatives.md [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+
+### <a href="https://www.django-rest-framework.org/" class="external-link" target="_blank">Django REST Framework</a>
+
+Django REST framework foi criado para ser uma caixa de ferramentas flexível para construção de APIs web utilizando Django por baixo, para melhorar suas capacidades de API.
+
+Ele é utilizado por muitas companhias incluindo Mozilla, Red Hat e Eventbrite.
+
+Ele foi um dos primeiros exemplos de **documentação automática de API**, e essa foi especificamente uma das primeiras idéias que inspirou "a busca por" **FastAPI**.
+
+!!! nota
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:33]:
```suggestion
!!! note
```

File: docs/pt/docs/alternatives.md [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+    Django REST Framework foi criado por Tom Christie. O mesmo criador de Starlette e Uvicorn, nos quais **FastAPI** é baseado.
+
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:37]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+
+Por ser tão simples, é relativamente intuitivo de aprender, embora a documentação esteja de forma mais técnica em alguns pontos.
+
+Ele é comumente utilizado por outras aplicações que não necessariamente precisam de banco de dados, gerenciamento de usuários, ou algum dos muitos recursos que já vem instalados no Django. Embora muitos desses recursos possam ser adicionados com plug-ins.
+
+Esse desacoplamento de partes, e sendo um "microframework" que pode ser extendido para cobrir exatamente o que é necessário era um recurso chave que eu queria manter.
+
+Dada a simplicidade do Flask, parecia uma ótima opção para construção de APIs. A próxima coisa a procurar era um "Django REST Framework" para Flask.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:54]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+```Python hl_lines="1"
+@app.get("/some/url")
+def read_url():
+    return {"message": "Hello World"}
+```
+
+Veja as similaridades em `requests.get(...)` e `@app.get(...)`.
+
+!!! checado "**FastAPI** isnpirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:94]:
```suggestion
!!! check "**FastAPI** isnpirado para"
```

File: docs/pt/docs/alternatives.md [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+
+Então eu descobri que existia um padrão para documentar APIs, utilizando JSON (ou YAML, uma extensão do JSON) chamado Swagger.
+
+E tinha uma interface web para APIs Swagger já criada. Então, sendo capaz de gerar documentação Swagger para uma API poderia permitir utilizar essa interface web automaticamente.
+
+Em algum ponto, Swagger foi dado para a Fundação Linux, e foi renomeado OpenAPI.
+
+Isso acontece porquê quando alguém fala sobre a versão 2.0 é comum dizer "Swagger", e para a versão 3+, "OpenAPI".
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:112]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+
+Outro grande recurso necessário nas APIs é validação de dados, certificando que os dados são válidos, dados certos parâmetros. Por exemplo, algum campo é `int`, e não alguma string aleatória. Isso é especialmente útil para dados que estão chegando.
+
+Sem um sistema de validação de dados, você teria que realizar todas as verificações manualmente, no código.
+
+Esses recursos são o que Marshmallow foi construído para fornecer. Ele é uma ótima biblioteca, e eu já utilizei muito antes.
+
+Mas ele foi criado antes da existência do _type hints_ do Python. Então, para definir todo o <abbr title="definição de como os dados devem ser formados">_schema_</abbr> você precisa utilizar específicas ferramentas e classes fornecidas pelo Marshmallow.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:138]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 7]
```diff
@@ -0,0 +1,417 @@
...
+Webargs é uma ferramente feita para fornecer o que está no topo de vários frameworks, inclusive Flask.
+
+Ele utiliza Marshmallow por baixo para validação de dados. E ele foi criado pelos mesmos desenvolvedores.
+
+Ele é uma grande ferramenta e eu também a utilizei muito, antes de ter o **FastAPI**.
+
+!!! info
+    Webargs foi criado pelos mesmos desenvolvedores do Marshmallow.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:154]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 8]
```diff
@@ -0,0 +1,417 @@
...
+
+Mas então, nós temos novamente o problema de ter uma micro-sintaxe, dentro de uma string Python (um grande YAML).
+
+O editor não poderá ajudar muito com isso. E se nós modificarmos os parâmetros dos _schemas_ do Marshmallow e esquecer de modificar também aquela _docstring_ YAML, o _schema_ gerado pode ficar obsoleto.
+
+!!! info
+    APISpec foi criado pelos mesmos desenvolvedores do Marshmallow.
+
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:179]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 9]
```diff
@@ -0,0 +1,417 @@
...
+* <a href="https://github.com/tiangolo/full-stack" class="external-link" target="_blank">https://github.com/tiangolo/full-stack</a>
+* <a href="https://github.com/tiangolo/full-stack-flask-couchbase" class="external-link" target="_blank">https://github.com/tiangolo/full-stack-flask-couchbase</a>
+* <a href="https://github.com/tiangolo/full-stack-flask-couchdb" class="external-link" target="_blank">https://github.com/tiangolo/full-stack-flask-couchdb</a>
+
+E esses mesmos geradores _full-stack_ foram a base do [ Projeto de Geradores **FastAPI**](project-generation.md){.internal-link target=_blank}.
+
+!!! info
+    Flask-apispec foi criado pelos mesmos desenvolvedores do Marshmallow.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:205]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 10]
```diff
@@ -0,0 +1,417 @@
...
+
+Ele tem um sistema de injeção de dependência integrado, inspirado pelo Angular dois. É necessário fazer o pré-registro dos "injetáveis" (como todos os sistemas de injeção de dependência que conheço), então, adicionando verbosidade e repetição de código.
+
+Como os parâmetros são descritos com tipos TypeScript (similar aos _type hints_ do Python), o suporte ao editor é muito bom.
+
+Mas como os dados TypeScript não são preservados após a compilação para o JavaScript, ele não pode depender dos tipos para definir a validação, serialização e documentação ao mesmo tempo. Devido a isso e a algumas decisões de projeto, para pegar a validação, serialização e geração automática do _schema_, é necessário adicionar decoradores em muitos lugares. Então, ele se torna muito verboso.
+
+Ele também não controla modelos aninhados muito bem. Então, se o corpo JSON na requisição for um objeto JSON que contém campos internos que contém objetos JSON aninhados, ele não consegue ser validado e documentado apropriadamente.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:222]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 11]
```diff
@@ -0,0 +1,417 @@
...
+    Usar tipos Python para ter um ótimo suporte do editor.
+
+    Ter um sistema de injeção de dependência poderoso. Achar um jeito de minimizar repetição de código.
+
+### <a href="https://sanic.readthedocs.io/en/latest/" class="external-link" target="_blank">Sanic</a>
+
+Ele foi um dos primeiros frameworks Python extremamente rápido baseado em `asyncio`. Ele foi feito para ser muito similar ao Flask.
+
+!!! nota "Detalhes técnicos"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:231]:
```suggestion
!!! note "Detalhes técnicos"
```

File: docs/pt/docs/alternatives.md [hunk 12]
```diff
@@ -0,0 +1,417 @@
...
+    Ele utiliza <a href="https://github.com/MagicStack/uvloop" class="external-link" target="_blank">`uvloop`</a> ao invés do '_loop_' `asyncio` padrão do Python. É isso que deixa ele tão rápido.
+
+    Ele claramente inspirou Uvicorn e Starlette, que são atualmente mais rápidos que o Sanic em testes de performance abertos.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:236]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+
+Falcon é outro framework Python de alta performance, e é projetado para ser minimalista, e funciona como fundação de outros frameworks como Hug.
+
+Ele usa o padrão anterior para frameworks web Python (WSGI) que é síncrono, então ele não pode controlar _WebSockets_ e outros casos de uso. No entanto, ele também tem uma boa performance.
+
+Ele é projetado para ter funções que recebem dois parâmetros, uma "requisição" e uma "resposta". Então você "lê" as partes da requisição, e "escreve" partes para a resposta. Devido ao seu design, não é possível declarar parâmetros de requisição e corpos com _type hints_ Python padrão como parâmetros de funções.
+
+Então, validação de dados, serialização e documentação tem que ser feitos no código, não automaticamente. Ou eles terão que ser implementados como um framework acima do Falcon, como o Hug. Essa mesma distinção acontece em outros frameworks que são inspirados pelo design do Falcon, tendo um objeto de requisição e um objeto de resposta como parâmetros.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:251]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 14]
```diff
@@ -0,0 +1,417 @@
...
+
+Ele não utiliza validação de dados, seriallização e documentação de bibliotecas de terceiros como o Pydantic, ele tem seu prórpio. Então, essas definições de tipo de dados não podem ser reutilizados tão facilmente.
+
+Ele exige um pouco mais de verbosidade nas configurações. E como é baseado no WSGI (ao invés de ASGI), ele não é projetado para ter a vantagem da alta performance fornecida por ferramentas como Uvicorn, Starlette e Sanic.
+
+O sistema de injeção de dependência exige pré-registro das dependências e as dependências são resolvidas baseadas nos tipos declarados. Então, não é possível declarar mais do que um "componente" que fornece um certo tipo.
+
+Rotas são declaradas em um único lugar, usando funções declaradas em outros lugares (ao invés de usar decoradores que possam ser colocados diretamente acima da função que controla o _endpoint_). Isso é mais perto de como o Django faz isso do que como Flask (e Starlette) faz. Ele separa no código coisas que são relativamente amarradas.
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:274]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 15]
```diff
@@ -0,0 +1,417 @@
...
+Ele não era baseado em um padrão como OpenAPI e JSON Schema. Então não poderia ter interação direta com outras ferramentas, como Swagger UI. Mas novamente, era uma idéia muito inovadora.
+
+Hug tinha um incomum, interessante recurso: usando o mesmo framework, é possível criar tanto APIs como CLIs.
+
+Como é baseado nos padrões anteriores de frameworks web síncronos (WSGI), ele não pode controlar _Websockets_ e outras coisas, embora ele ainda tenha uma alta performance também.
+
+!!! info
+    Hug foi criado por Timothy Crosley, o mesmo criador do  <a href="https://github.com/timothycrosley/isort" class="external-link" target="_blank">`isort`</a>, uma grande ferramenta para ordenação automática de _imports_ em arquivos Python.
+
+!!! checado "Idéias inspiradas para o **FastAPI**"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:296]:
```suggestion
!!! check "Idéias inspiradas para o **FastAPI**"
```

File: docs/pt/docs/alternatives.md [hunk 16]
```diff
@@ -0,0 +1,417 @@
...
+Agora APIStar é um conjunto de ferramentas para validar especificações OpenAPI, não um framework web.
+
+!!! info
+    APIStar foi criado por Tom Christie. O mesmo cara que criou:
+
+    * Django REST Framework
+    * Starlette (no qual **FastAPI** é baseado)
+    * Uvicorn (usado por Starlette e **FastAPI**)
+
+!!! checado "**FastAPI** inspirado para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:334]:
```suggestion
!!! check "**FastAPI** inspirado para"
```

File: docs/pt/docs/alternatives.md [hunk 17]
```diff
@@ -0,0 +1,417 @@
...
+
+### <a href="https://pydantic-docs.helpmanual.io/" class="external-link" target="_blank">Pydantic</a>
+
+Pydantic é uma biblioteca para definir validação de dados, serialização e documentação (usando JSON Schema) baseado nos Python _type hints_.
+
+Isso faz dele extremamente intuitivo.
+
+Ele é comparável ao Marshmallow. Embora ele seja mais rápido que Marshmallow em testes de performance. E ele é baseado nos mesmos Python _type hints_, o suporte ao editor é ótimo.
+
+!!! checado "**FastAPI** usa isso para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:355]:
```suggestion
!!! check "**FastAPI** usa isso para"
```

File: docs/pt/docs/alternatives.md [hunk 18]
```diff
@@ -0,0 +1,417 @@
...
+
+Starlette é atualmente o mais rápido framework Python testado. Somente ultrapassado pelo Uvicorn, que não é um framework, mas um servidor.
+
+Starlette fornece toda a funcionalidade básica de um microframework web.
+
+Mas ele não fornece validação de dados automática, serialização e documentação.
+
+Essa é uma das principais coisas que **FastAPI** adiciona no topo, tudo baseado em Python _type hints_ (usando Pydantic). Isso, mais o sistema de injeção de dependência, utilidades de segurança, geração de _schema_ OpenAPI etc.
+
+!!! nota "Detalhes Técnicos"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:388]:
```suggestion
!!! note "Detalhes Técnicos"
```

File: docs/pt/docs/alternatives.md [hunk 19]
```diff
@@ -0,0 +1,417 @@
...
+    ASGI é um novo "padrão" sendo desenvolvido pelos membros do time central do Django. Ele ainda não está como "Padrão Python" (PEP), embora eles estejam em processo de fazer isso.
+
+    No entanto, ele já está sendo utilizado como "padrão" por diversas ferramentas. Isso melhora enormemente a interoperabilidade, como você poderia trocar Uvicorn por qualquer outro servidor ASGI (como Daphne ou Hypercorn), ou você poderia adicionar ferramentas compatíveis com ASGI, como `python-socketio`.
+
+!!! checado "**FastAPI** usa isso para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:393]:
```suggestion
!!! check "**FastAPI** usa isso para"
```

File: docs/pt/docs/alternatives.md [hunk 20]
```diff
@@ -0,0 +1,417 @@
...
+
+### <a href="https://www.uvicorn.org/" class="external-link" target="_blank">Uvicorn</a>
+
+Uvicorn é um servidor ASGI peso leve, construído com uvloop e httptools.
+
+Ele não é um framework web, mas sim um servidor. Por exemplo, ele não fornece ferramentas para roteamento por rotas. Isso é algo que um framework como Starlette (ou **FastAPI**) poderia fornecer por cima.
+
+Ele é o servidor recomendado para Starlette e **FastAPI**.
+
+!!! checado "**FastAPI** recomenda isso para"
```
Comment (cassiobotaro) [inline docs/pt/docs/alternatives.md:408]:
```suggestion
!!! check "**FastAPI** recomenda isso para"
```

**Review by cassiobotaro — CHANGES_REQUESTED** [review]
minor NIT's

**Review by cassiobotaro — APPROVED** [review]

**Comment by cassiobotaro** [comment]
Thanks @Serrones !

**Comment by tiangolo** [comment]
Awesome, thank you @Serrones ! :rocket: :tada: 

I just added a couple of small format and wording changes :heavy_check_mark: 

And thanks @cassiobotaro for the review! :nerd_face: :coffee:

**Pull request merged**